### ISSUE TITLE
rflash slow to respond and two operations can hang

### ISSUE THREAD

**Issue by marksmit-ibm** [xcat2/xcat-core#5276]
Testing issue https://github.com/xcat2/xcat-core/pull/5274

I see getting ready to run an update, I first run 
rflash service -l    (service = group of sn01,sn02)  
It takes 43 seconds real time to complete but almost no time to run:
(0) root @ xcat1: /firmware/Witherspoon/OP910/OP910_1742FxR8-910.1806.20180530t-pnorOnlyUpdate
# time rflash service witherspoon.pnor.squashfs.tar -c 
TAR xyz.openbmc_project.Software.Version.VersionPurpose.Host Firmware Product Version: IBM-witherspoon-ibm-OP9_v1.19_1.168
...
real    0m43.583s  <== seems too long compared to 2.14.1-May-05 version running previously.
user    0m0.128s
sys 0m0.010s

Next I proceed to test issue 5274 by starting `rflash service <pnor file> -a`
in one window and then in 2nd xterm, I start `rflash service -l`

# ps -ef |grep rflash
root      46251  89558  0 18:52 pts/46   00:00:00 perl /opt/xcat/bin/rflash service witherspoon.pnor.squashfs.tar -a
root      46255 106816  0 18:52 ?        00:00:00 xcatd SSL: rflash to service for root@localhost
root      46257  46255  0 18:52 ?        00:00:00 xcatd SSL: rflash to service for root@localhost: openbmc instance
root      46603  46257  0 18:52 ?        00:00:00 xcatd SSL: rflash to service for root@localhost: openbmc instance
root      50522  48601  0 18:57 pts/18   00:00:00 grep --color=auto rflash
(0) root @ xcat1: /root
# strace -p 46257
...
waitpid(-1, 0x7fffc24ba3c4, WNOHANG)    = 0
_newselect(0, NULL, NULL, NULL, {0, 10000}) = 0 (Timeout)
waitpid(-1, 0x7fffc24ba3c4, WNOHANG)    = 0
_newselect(0, NULL, NULL, NULL, {0, 10000}) = 0 (Timeout)
waitpid(-1, 0x7fffc24ba3c4, WNOHANG)    = 0
_newselect(0, NULL, NULL, NULL, {0, 10000}) = 0 (Timeout)
waitpid(-1, 0x7fffc24ba3c4, WNOHANG)    = 0
_newselect(0, NULL, NULL, NULL, {0, 10000}) = 0 (Timeout)
waitpid(-1, 0x7fffc24ba3c4, WNOHANG)    = 0
_newselect(0, NULL, NULL, NULL, {0, 10000}) = 0 (Timeout)
waitpid(-1, 0x7fffc24ba3c4, WNOHANG)    = 0
_newselect(0, NULL, NULL, NULL, {0, 10000}) = 0 (Timeout)
waitpid(-1, 0x7fffc24ba3c4, WNOHANG)    = 0
_newselect(0, NULL, NULL, NULL, {0, 10000}) = 0 (Timeout)
waitpid(-1, 0x7fffc24ba3c4, WNOHANG)    = 0
_newselect(0, NULL, NULL, NULL, {0, 10000}) = 0 (Timeout)
waitpid(-1, 0x7fffc24ba3c4, WNOHANG)    = 0
_newselect(0, NULL, NULL, NULL, {0, 10000}) = 0 (Timeout)
waitpid(-1, 0x7fffc24ba3c4, WNOHANG)    = 0

After some time, `rflash service -l` does complete.  I restarted the command for time.
real	0m36.595s
user	0m0.126s
sys	0m0.009s

# ps -ef |grep rflash
root      46251  89558  0 18:52 pts/46   00:00:00 perl /opt/xcat/bin/rflash service witherspoon.pnor.squashfs.tar -a
root      46255 106816  0 18:52 ?        00:00:00 xcatd SSL: rflash to service for root@localhost
root      46257  46255  0 18:52 ?        00:00:02 xcatd SSL: rflash to service for root@localhost: openbmc instance
root      46603  46257  0 18:52 ?        00:00:00 xcatd SSL: rflash to service for root@localhost: openbmc instance
root      60081  16912  0 19:07 pts/13   00:00:00 perl /opt/xcat/bin/rflash service -l
root      60086 106816  0 19:07 ?        00:00:00 xcatd SSL: rflash to service for root@localhost
root      60088  60086  0 19:07 ?        00:00:00 xcatd SSL: rflash to service for root@localhost: openbmc instance

Meanwhile `rflash service -a` is still running:
# ps -ef |grep rflash
root      46251  89558  0 18:52 pts/46   00:00:00 perl /opt/xcat/bin/rflash service witherspoon.pnor.squashfs.tar -a
root      46255 106816  0 18:52 ?        00:00:00 xcatd SSL: rflash to service for root@localhost
root      46257  46255  0 18:52 ?        00:00:03 xcatd SSL: rflash to service for root@localhost: openbmc instance
root      46603  46257  0 18:52 ?        00:00:00 xcatd SSL: rflash to service for root@localhost: openbmc instance

# strace -p 46603
strace: Process 46603 attached
read(14, <<== stops here.  After 20 seconds I control-c.  
^Cstrace: Process 46603 detached
 <detached ...>

**Comment by whowutwut** [comment]
@xuweibj @robin2008 can you take a look here?

**Comment by robin2008** [comment]
@xuweibj 
I think we need lock for those actions which cannot be applied at the same time, planned in 2.14.2

**Comment by robin2008** [comment]
It might not be related to our codes,  I just do some investigation on Mark's environment.  And found that the `uploading` cannot be successfully.
```
# tail /var/log/xcat/rflash/sn02.log
Activating Firmware ibm-v2.0-0-r46-0-gbed584c...  30%
Activating Firmware ibm-v2.0-0-r46-0-gbed584c...  30%
Activating Firmware ibm-v2.0-0-r46-0-gbed584c...  30%
Activating Firmware ibm-v2.0-0-r46-0-gbed584c...  30%
sn02: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.167 activation successful.
Uploading /firmware/Witherspoon/OP910/OP910_1742FxR8-910.1806.20180530t-pnorOnlyUpdate/witherspoon.pnor.squashfs.tar ...
Uploading /firmware/Witherspoon/OP910/OP910_1742FxR8-910.1806.20180530t-pnorOnlyUpdate/witherspoon.pnor.squashfs.tar ...
Uploading /firmware/Witherspoon/OP910/OP910_1742FxR8-910.1806.20180530t-pnorOnlyUpdate/witherspoon.pnor.squashfs.tar ...
Uploading /firmware/Witherspoon/OP910/OP910_1742FxR8-910.1806.20180530t-pnorOnlyUpdate/witherspoon.pnor.squashfs.tar ...
Uploading /firmware/Witherspoon/OP910/OP910_1742FxR8-910.1806.20180530t-pnorOnlyUpdate/witherspoon.pnor.squashfs.tar ...
```

```
# ps -ef|grep curl
root      58985  47830  0 23:55 pts/28   00:00:00 grep --color=auto curl
root     130272 130054  0 23:07 ?        00:00:01 curl -b /tmp/_xcat_cjar.sn02 -k -H Content-Type: application/octet-stream -X PUT -T /firmware/Witherspoon/OP910/OP910_1742FxR8-910.1806.20180530t-pnorOnlyUpdate/witherspoon.pnor.squashfs.tar https://192.168.85.3/upload/image/
```
And from the log, we can see that after `IBM-witherspoon-ibm-OP9_v1.19_1.167 ` is activated, the uploading is not finished again.

**Comment by marksmit-ibm** [comment]
With our version of xCAT unchanged, I rebooted sn02-bmc 2 times until `rpower sn02 stat` and `rflash sn02 -l` again properly returned without error.
I consulted with openbmc developer and confirmed that:
1. the upload should extract to /tmp, 
2. the "rwfs" filesystems are 89% so wouldn't think they are full
3. check /tmp/images/ which is where the file is copied then extracted

I then attempted recreate with one instance of  -a, thus:
```
> # rflash sn01,sn02 -l
sn02: ID       Purpose State      Version
sn02: -------------------------------------------------------
sn02: 7c6bacc0 Host    Active(*)  IBM-witherspoon-ibm-OP9_v1.19_1.167
sn02: 369a7049 Host    Active     IBM-witherspoon-ibm-OP9-v2.0-2.14
sn02: b04fff27 BMC     Active(*)  ibm-v2.0-0-r46-0-gbed584c
sn02: 6017da28 BMC     Active     ibm-v2.1-438-g0030304-r12-0-g5ee4fb0
sn02: 
sn01: ID       Purpose State      Version
sn01: -------------------------------------------------------
sn01: 7c6bacc0 Host    Active     IBM-witherspoon-ibm-OP9_v1.19_1.167
sn01: 2ca7e103 Host    Active(*)  IBM-witherspoon-ibm-OP9_v1.19_1.168
sn01: b04fff27 BMC     Active(*)  ibm-v2.0-0-r46-0-gbed584c
sn01: 6017da28 BMC     Active     ibm-v2.1-438-g0030304-r12-0-g5ee4fb0
sn01: 

(0) root @ xcat1: /firmware/Witherspoon/OP910/OP910_1742FxR8-910.1806.20180530t-pnorOnlyUpdate   <== Directory of 1.168 pnor level file

> # rflash sn01,sn02 witherspoon.pnor.squashfs.tar -a
Attempting to upload /firmware/Witherspoon/OP910/OP910_1742FxR8-910.1806.20180530t-pnorOnlyUpdate/witherspoon.pnor.squashfs.tar, please wait...
sn02: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.168 (ID: 2ca7e103)
sn01: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.168 (ID: 2ca7e103)
sn01: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.168 activation successful.
```
^^^ (above returned within 30 seconds, which is good.)  Remember sn01 is already running 1.168 level pnor, so I would expect it to complete quickly (nothing to change).
At this point I ran in a 2nd xterm, and got a new message of progress (not seen before) of "Activating".  It took awhile to return but did return:
```
> # ps -ef |grep rflash
root     110918  89558  0 09:16 pts/46   00:00:00 perl /opt/xcat/bin/rflash sn01,sn02 witherspoon.pnor.squashfs.tar -a
root     110922 106816  0 09:16 ?        00:00:00 xcatd SSL: rflash to sn01,sn02 for root@localhost
root     110924 110922  4 09:16 ?        00:00:02 xcatd SSL: rflash to sn01,sn02 for root@localhost: openbmc instance

> # rflash sn02 -l
sn02: ID       Purpose State      Version
sn02: -------------------------------------------------------
sn02: 7c6bacc0 Host    Active(*)  IBM-witherspoon-ibm-OP9_v1.19_1.167
sn02: 2ca7e103 Host    Activating IBM-witherspoon-ibm-OP9_v1.19_1.168  <<== Looks ok
sn02: b04fff27 BMC     Active(*)  ibm-v2.0-0-r46-0-gbed584c
sn02: 6017da28 BMC     Active     ibm-v2.1-438-g0030304-r12-0-g5ee4fb0
sn02: 

(0) root @ xcat1: /root
# ps -ef |grep rflash
root     110918  89558  0 09:16 pts/46   00:00:00 perl /opt/xcat/bin/rflash sn01,sn02 witherspoon.pnor.squashfs.tar -a
root     110922 106816  0 09:16 ?        00:00:00 xcatd SSL: rflash to sn01,sn02 for root@localhost
root     110924 110922 33 09:16 ?        00:01:10 xcatd SSL: rflash to sn01,sn02 for root@localhost: openbmc instance
```
sn02: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.168 activation successful.
-------------------------------------------------------
Firmware update complete: Total=2 Success=2 Failed=0
-------------------------------------------------------

In summary, repeating the test with just one instance of 'rflash -a' plus injecting one instance of 'rflash sn02 -l'  successfully completed.
I believe only my original scenario, where two instances of 'rflash -a' are accidentally run, along with an instance of 'rflash -l' will recreate this issue.

**Comment by robin2008** [comment]
Thanks @marksmit-ibm 

And @xuweibj   @gurevichmark 
Do we have lock to prevent run the `rflash -a` for same node ?  For example,  in one terminal, run
`rflash cn1,cn2  ... -a` and in the same time to open another terminal , and run `rflash cn2 ... -a`

**Comment by robin2008** [comment]
@xuweibj  Could you do more testing for it?  Need answer below question:
1, Does the BMC will run into trouble if admin issue two `rflash` in the same time?

If yes, then we have to work out a lock method to avoid this.  Maybe we can refer to the ipmi implementation.

**Comment by xuweibj** [comment]
My test scenarios as below:
2 Terminals to run rflash command 
* 1 Run ``rflash -u`` on 2 terminals at the same time, it seems after one upload success the other one start to upload. Final, success.
* 2 After one terminal shows that "Uploading **", run ``rflash -u`` on another terminal. After first one get "successful", the second one shows "Uploading". Final, success.
* 3 One terminal run ``rflash -a`` first, when shows "upload success" run ``rflash -u`` on the other one, second one upload again with the first one keeps activating status.
* 4 One terminal  run ``rflash -a`` first, when shows "upload success" run ``rflash -a`` on the other one, second one upload again with the first one keeps activating status. And final all get "activation successful".
* 5 Run ``rflash -a`` at the same time, success.

For scenario 2, even encountered ``Error: [500] Login to BMC failed: 500 Can't connect to 172.11.139.16:443.`` on the second terminal. But did not affect upload.

**Comment by gurevichmark** [comment]
Tried `rflash <node> -a` one after the other with 3 seconds apart to the same node. First activation fails, but the second succeeds. I think we need some kind of lock to prevent second upload/activation from starting:
```
[root@briggs01 gurevich]#  rflash mid05tor12cn13 -a /mnt/xcat/iso/openbmc/op920/920.1810.20180515n/obmc-witherspoon-ibm-v2.1.ubi.mtd.tar -V& sleep 3;rflash mid05tor12cn13 -a /mnt/xcat/iso/openbmc/op920/920.1824.20180608a/obmc-witherspoon-ibm-v2.1.ubi.mtd.tar -V&
[1] 107835
[briggs01]: Attempting to upload /mnt/xcat/iso/openbmc/op920/920.1810.20180515n/obmc-witherspoon-ibm-v2.1.ubi.mtd.tar, please wait...
[briggs01]: Running command in Perl
[2] 107871
[root@briggs01 gurevich]# mid05tor12cn13: [briggs01]: DD 2.1 processor
mid05tor12cn13: [briggs01]: Uploading /mnt/xcat/iso/openbmc/op920/920.1810.20180515n/obmc-witherspoon-ibm-v2.1.ubi.mtd.tar ...
[briggs01]: Attempting to upload /mnt/xcat/iso/openbmc/op920/920.1824.20180608a/obmc-witherspoon-ibm-v2.1.ubi.mtd.tar, please wait...
[briggs01]: Running command in Perl
mid05tor12cn13: [briggs01]: Firmware upload successful. Attempting to activate firmware: ibm-v2.1-438-g0030304-r11-0-g80e9668 (ID: 718dbc93)
mid05tor12cn13: [briggs01]: rflash ibm-v2.1-438-g0030304-r11-0-g80e9668 started, please wait...
mid05tor12cn13: [briggs01]: DD 2.1 processor
mid05tor12cn13: [briggs01]: Uploading /mnt/xcat/iso/openbmc/op920/920.1824.20180608a/obmc-witherspoon-ibm-v2.1.ubi.mtd.tar ...
mid05tor12cn13: [briggs01]: Activating Firmware ibm-v2.1-438-g0030304-r11-0-g80e9668...  30%
mid05tor12cn13: [briggs01]: Activating Firmware ibm-v2.1-438-g0030304-r11-0-g80e9668...  30%
mid05tor12cn13: [briggs01]: Firmware upload successful. Attempting to activate firmware: ibm-v2.1-438-g0030304-r15-0-g19832d3 (ID: 376af621)
mid05tor12cn13: [briggs01]: rflash ibm-v2.1-438-g0030304-r15-0-g19832d3 started, please wait...
mid05tor12cn13: [briggs01]: Error: Firmware ibm-v2.1-438-g0030304-r15-0-g19832d3 activation failed.
[briggs01]: -------------------------------------------------------
[briggs01]: Firmware update complete: Total=1 Success=0 Failed=1
[briggs01]: mid05tor12cn13: Firmware ibm-v2.1-438-g0030304-r15-0-g19832d3 activation failed.
[briggs01]: -------------------------------------------------------
mid05tor12cn13: [briggs01]: Activating Firmware ibm-v2.1-438-g0030304-r11-0-g80e9668...  30%
mid05tor12cn13: [briggs01]: Activating Firmware ibm-v2.1-438-g0030304-r11-0-g80e9668...  30%
mid05tor12cn13: [briggs01]: Activating Firmware ibm-v2.1-438-g0030304-r11-0-g80e9668...  30%
mid05tor12cn13: [briggs01]: Activating Firmware ibm-v2.1-438-g0030304-r11-0-g80e9668...  30%
mid05tor12cn13: [briggs01]: Activating Firmware ibm-v2.1-438-g0030304-r11-0-g80e9668...  30%
mid05tor12cn13: [briggs01]: Activating Firmware ibm-v2.1-438-g0030304-r11-0-g80e9668...  30%
mid05tor12cn13: [briggs01]: Activating Firmware ibm-v2.1-438-g0030304-r11-0-g80e9668...  30%
mid05tor12cn13: [briggs01]: Activating Firmware ibm-v2.1-438-g0030304-r11-0-g80e9668...  30%
mid05tor12cn13: [briggs01]: Activating Firmware ibm-v2.1-438-g0030304-r11-0-g80e9668...  30%
[briggs01]: mid05tor12cn13: Firmware ibm-v2.1-438-g0030304-r11-0-g80e9668 activation successful.
[briggs01]: -------------------------------------------------------
[briggs01]: Firmware update complete: Total=1 Success=1 Failed=0
[briggs01]: -------------------------------------------------------

[1]-  Done                    rflash mid05tor12cn13 -a /mnt/xcat/iso/openbmc/op920/920.1810.20180515n/obmc-witherspoon-ibm-v2.1.ubi.mtd.tar -V
[2]+  Exit 1                  rflash mid05tor12cn13 -a /mnt/xcat/iso/openbmc/op920/920.1824.20180608a/obmc-witherspoon-ibm-v2.1.ubi.mtd.tar -V
[root@briggs01 gurevich]#
``` 

```
[root@briggs01 gurevich]# rflash mid05tor12cn13 -l
mid05tor12cn13: ID       Purpose State      Version
mid05tor12cn13: -------------------------------------------------------
mid05tor12cn13: d8368f51 Host    Active(*)  IBM-witherspoon-ibm-OP9_v1.19_1.172
mid05tor12cn13: b04fff27 BMC     Active(*)  ibm-v2.0-0-r46-0-gbed584c
mid05tor12cn13: 376af621 BMC     Failed     ibm-v2.1-438-g0030304-r15-0-g19832d3
mid05tor12cn13: 49b8ef1f Host    Active     IBM-witherspoon-ibm-OP9_v1.19_1.149
mid05tor12cn13: 7a1ce9fe BMC     Active     ibm-v2.0-0-r43-0-gdb0b7b3
mid05tor12cn13: 718dbc93 BMC     Active(+)  ibm-v2.1-438-g0030304-r11-0-g80e9668
mid05tor12cn13:
[root@briggs01 gurevich]#
```

**Comment by xuweibj** [comment]
I think we could create ``<node>-rflash.lock`` lock file for each node under ``/var/lock/xcat/`` when command is ``rflash -u/-a/-d``. 
Check whether has lock file for itself, if has, print out msg with pid and exit. If no, continue doing rflash.
After end of each node rflash, delete lock file.

**Comment by chuckbrazie** [comment]
I am out of the office until 07/16/2018.

I am out of the office until  Monday July 16.


Note: This is an automated response to your message  "Re: [xcat2/xcat-core]
rflash slow to respond and two operations can hang (#5276)" sent on 7/2/18
8:11:21 AM.

This is the only notification you will receive while this person is away.

**Comment by gurevichmark** [comment]
@xuweibj Yes, that is exactly my plan. I am working on a PR that would use `Utils->acquire_lock()` to create `<node>-rflash.lock` file when upload or activate action is being processed, then `Utils->release_lock()` at the end. The difficulty is finding all the places where `rflash` exits so that the lock is released. If we miss some exit path and not release the lock, future `rflash` commands will not go through.

**Comment by chuckbrazie** [comment]
I am out of the office until 07/16/2018.

I am out of the office until  Monday July 16.


Note: This is an automated response to your message  "Re: [xcat2/xcat-core]
rflash slow to respond and two operations can hang (#5276)" sent on 7/2/18
1:10:24 PM.

This is the only notification you will receive while this person is away.
